chart - added gnuplot logscale support

dragazo opened the pull request:
This PR adds support for the `logscale` option of gnuplot to the `chart` service. Usage-wise, this is another option called `logscale`, which expects either a string (axes), or a list with string (axes) and number (log base). I decided the [gnuplot syntax](http://gnuplot.sourceforge.net/docs_4.2/node197.html) for specifying axes was simple enough, so this just uses that directly. So if you want to scale the y axis you can just do `['logscale', 'y']` for base 10, or `['logscale', ['y', 2]]` for base 2.

brollb reviewed (changes requested):
Nice work. I added a few comments. It would also be great if you could add some more tests. I refactored the Chart service a little to make it easier to test the options parsing logic. Could you add a couple more tests?

dragazo commented on src/server/services/procedures/chart/chart.js:
```diff
@@ -207,6 +208,22 @@ chart.draw = function(lines, options={}){
+        else if (Array.isArray(val)) {
+            const axes = val[0];
+            let base = val[1];
+            if (typeof axes !== 'string') throw Error('logscale axes was not a string (text)');
```
From the gnuplot reference, it can actually be "any combination of `x`, `x2`, `y`, `y2`, `z`, and `cb` in any order". But I could add some regex to check that.

dragazo replied:
Oh, and for the if statement for string type, the only time I've seen that be invoked is if you pass it a list.

brollb replied:
Do you have an example where we might use some of the other options? If they aren't pressing for us, I would opt for supporting just "x" and "y". This should make the error messages easier to understand, too.

dragazo commented on src/server/services/procedures/chart/chart.js:
```diff
@@ -207,6 +208,22 @@ chart.draw = function(lines, options={}){
+            if (typeof axes !== 'string') throw Error('logscale axes was not a string (text)');
+            if (base !== undefined) {
+                base = +base;
+                if (!base) throw Error('logscale base was not a number'); // not strictly necessary for gnuplot, but nice for users
```
Yeah, I initially allowed negative since gnuplot seems to just take the absolute value internally, but it's probably better to forbid it anyway. Combined with the 'not a number' logic to make the error message clear in all cases (I hadn't noticed before that js makes `+''===0`)

brollb replied:
Yeah, I agree that it would make more sense to simply disallow nonsensical values (even if allowed by gnuplot). I would also recommend just using BoundedNumber<2> from the input-types.js file so error messages are consistent and we are reusing type parsing code (rather than reinventing the wheel).

dragazo commented:
Ok, I added some more tests for usage and made the changes you mentioned for the input validation. I strengthened the axes string validation to make sure that it's a string of legal tokens (x2, y2, x, y, z, cb), but it doesn't forbid redundancies (I could add that if we want, but gnuplot is fine with duplicates).

brollb commented on test/unit/server/services/procedures/chart/chart.spec.js:
```diff
@@ -32,11 +32,65 @@ describe(utils.suiteName(__filename), function() {
+            const [, options] = Chart._parseDrawInputs([], {logscale: ''});
+            assert.equal(options.logscale, undefined);
+        });
+        it('should allow empty axes', function(){
```
Duplicate test name

brollb replied:
Also, do we want to allow an empty list as the logscale? I am not sure this is the desired behavior.

dragazo replied:
I allowed it so that a "combine <list> using join(_,_)" block with a (potentially-empty) list of individual axis names would still be allowed (as a no-op). But I guess given how options work they could just conditionally add the logscale option itself.

brollb commented on test/unit/server/services/procedures/chart/chart.spec.js:
```diff
@@ -32,11 +32,65 @@ describe(utils.suiteName(__filename), function() {
+            assert.equal(options.logscale.axes, 'x');
+            assert.equal(options.logscale.base, 10);
+        });
+        it('should parse string logscale (using base 10)', function(){
```
duplicate test names

dragazo replied:
Oh, ok, I was treating them more like `expect` messages in rust tests. In that case, I'll merge the related ones together.

brollb commented on src/server/services/procedures/chart/chart.js:
```diff
@@ -207,22 +206,51 @@ chart.draw = function(lines, options={}){
+            if (typeof axes !== 'string') throw Error('logscale axes was not a string (text)');
+            if (base !== undefined) {
+                const temp = +base;
+                if (Array.isArray(base) || !temp || temp <= 0) throw Error('Invalid logscale option: base must be a positive number');
```
Why not use `BoundedNumber<1>` from [here](https://github.com/NetsBlox/NetsBlox/blob/master/src/server/services/input-types.js#L30) rather than reinventing the wheel?

It would look something like:
```javascript
try {
    base = InputTypes.parse.BoundedNumber(base, [1]);
} catch (err) {
    throw new Error(`Invalid logscale option: ${err.message}`);
}
```

dragazo replied:
Sure, that'll behave more consistently too. I actually didn't think to use those outside of the jsdoc spec.

dragazo commented:
The new commit addresses everything except the empty axes string, but if we indeed don't want to allow it, I can change that. I've left the axes validation as allowing all the x, y, x2, y2, z, cb that gnuplot allows, but I've simplified the error message to just mention x and y as examples.

brollb commented:
@dragazo - Can you take a look at the code climate issues? Remaining issues around complexity are OK as are code duplication (though should be reduced if possible) but others should generally be fixed.

dragazo commented:
Sure, I just fixed the code climate issues (now it just complains about slightly high complexity). I refactored a few pieces out of the logic, including the new `processOptions` function, which we could use promote to a general utility if other services need similar options lists.

brollb commented:
LGTM. Nice work, @dragazo!
